horovod/common/ops/mpi_operations.h

```diff
@@ -0,0 +1,161 @@
+// Copyright 2016 The TensorFlow Authors. All Rights Reserved.
+// Modifications copyright (C) 2019 Uber Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// =============================================================================
+
+#ifndef HOROVOD_MPI_OPERATIONS_H
+#define HOROVOD_MPI_OPERATIONS_H
+
+#include <iostream>
+
+#include "mpi.h"
+
+#include "../common.h"
+#include "../global_state.h"
+#include "collective_operations.h"
+
+#if HAVE_CUDA
+#include "cuda_operations.h"
+#endif
+
+namespace horovod {
+namespace common {
+
+struct MPIContext {
```

Comment:
I feel like `MPIContext` should be defined somewhere like `common.h`, since it is being used in the control plane as well. Might be more intuitive to have `mpi_operations.cc` and `mpi_operations.h` pertain only to the ops themselves. Also, the `Communicator` enum is defined there, so I think it would make more sense to have all the MPI communicator-related code in one place so that it is easier to try out custom communicators while developing.

Reply:
Good idea!  The MPIContext is a more generic object used in more places (like NCCL), so it makes sense to move it out of there.  I put it into a new file `mpi_context.h`.  In my mind, it makes sense to avoid putting it in `common.h`, since there is interest in decoupling Horovod from MPI in the future.  With this separation, we keep Horovod concepts independent of MPI.  What do you think?

Reply:
Agreed!